Serialization and Metadata API changes

PradyumnaKrishna opened the pull request:
### Description of the changes being introduced by the pull request:

BaseDeserializer requires a class type to deserialize the raw_data
into a class. This creates a problem with Paylaod Deserialization,
where we have to deserialize the same payload multiple time.

BaseDeserializer, JSONDeserializer returns json and that get
converted into the class object after that. Payload Deserialization
will be app specific and will be implemented at project level.

sign is renamed to create_sig and verify is renamed to verify_sigs,
so it does not clashes with other methods of current signature
wrappers.

### Please verify and check that the pull request fulfils the following requirements:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

PradyumnaKrishna committed:
Changes in serialization API

BaseDeserializer requires a class type to deserialize the raw_data
into a class. This creates a problem with Paylaod Deserialization,
where we have to deserialize the same payload multiple time.

BaseDeserializer, JSONDeserializer returns json and that get
converted into the class object after that. Payload Deserialization
will be app specific and will be implemented at project level.

Signed-off-by: Pradyumna Krishna <git@onpy.in>

PradyumnaKrishna committed:
Rename Envelope sign and verify methods

sign is renamed to create_sig and verify is renamed to verify_sigs,
so it does not clashes with other methods of current signature
wrappers.

This has to be done to avoid any breaking change while integrating
DSSE in in-toto and tuf projects.

Signed-off-by: Pradyumna Krishna <git@onpy.in>

PradyumnaKrishna committed:
Fix failing tests

Tests was failing due to changes in serialization and metadata
module. Some methods of DSSE Envelope was renamed which was
failing the test cases.

Signed-off-by: Pradyumna Krishna <git@onpy.in>

lukpueh commented on securesystemslib/metadata.py:
```diff
@@ -52,6 +52,16 @@ def default_deserializer() -> BaseDeserializer:
+    ) -> "Envelope":
+
+        json = super().from_bytes(data, deserializer)
+        return cls.from_dict(json)
```
I think you are mixing concerns here. Shouldn't we expect the deserializer passed to `from_bytes` to return an `Envelope`? IMO it's better and more flexible to move the logic here to a custom:

```python
class EnvelopeJSONDeserializer: # could also inherit from `JsonDeserializer` and call `super()` below
    def deserialize(data):
        return Envelope.from_dict(
            JsonDeserializer().deserialize(data)
        )
```

Then you can return that from `default_deserializer()` and it will provide you with the `from_bytes()` you want without the need to override the one inherited from `SerializationMixin`

Note that in-toto won't use this `EnvelopeJSONDeserializer` because there we want something like an `AnyMetadataJSONDeserializer`, which can return either an `Envelope` or `Metablock` as described in ITE-5. 

So for the purpose of ITE-5, you don't even have to create a `EnvelopeJSONDeserializer`. But I think it's okay to provide one in `securesystemslib` as blueprint for custom deserializers.

What do you think?

lukpueh commented on securesystemslib/serialization.py:
```diff
@@ -21,7 +21,7 @@ class BaseDeserializer(metaclass=abc.ABCMeta):
     @abc.abstractmethod
-    def deserialize(self, raw_data: bytes, cls: Any) -> Any:
+    def deserialize(self, raw_data: bytes) -> Any:
         """Deserialize bytes to a specific object."""
```
```suggestion
        """Deserialize bytes."""
```

lukpueh commented on securesystemslib/serialization.py:
```diff
@@ -30,26 +30,22 @@ def deserialize(self, raw_data: bytes, cls: Any) -> Any:
 
-    def deserialize(self, raw_data: bytes, cls: "JSONSerializable") -> Any:
-        """Deserialize utf-8 encoded JSON bytes into an instance of cls.
+    def deserialize(self, raw_data: bytes) -> Any:
```
```suggestion
    def deserialize(self, raw_data: bytes) -> Dict:
```
needs to be imported from typing

lukpueh commented on securesystemslib/serialization.py:
```diff
@@ -30,26 +30,22 @@ def deserialize(self, raw_data: bytes, cls: Any) -> Any:
-            cls: A "JSONSerializable" subclass.
 
         Returns:
             Object of the provided class type.
```
```suggestion
            dict
```

Maybe the Returns section of the docstring can be omitted, given the type hint and oneliner on top.

lukpueh commented on securesystemslib/serialization.py:
```diff
@@ -229,20 +225,10 @@ def to_file(
-    deserialization.
+    to dict. It provides `to_dict` method, therefore, JSONSerializer requires
+    a subclass of this class for serialization.
     """
```
Could be shortened to something like...

```python
    """Objects serialized with ``JSONSerializer`` must inherit from this class and
    implement its ``to_dict`` method.
    """
```

lukpueh reviewed (changes requested):
Great work!

PradyumnaKrishna committed:
Add EnvelopeJSONDeserializer

EnvelopeJSONDeserializer is a subclass of JSONDeserializer which
deserializes raw data into an object of Envelope class. This
deserializer is exclusive for DSSE Envelope.

To avoid circular imports EnvelopeJSONDeserializer is created in
metadata.py.

Signed-off-by: Pradyumna Krishna <git@onpy.in>

lukpueh commented on securesystemslib/serialization.py:
```diff
@@ -224,9 +227,8 @@ def to_file(
-    to dict. It provides `to_dict` method, therefore, JSONSerializer requires
-    a subclass of this class for serialization.
+    """Objects serialized with ``JSONSerializer`` must inherit from this class
+    and implement its ``to_dict`` method..
```
Nit pick:
```suggestion
    and implement its ``to_dict`` method.
```

PradyumnaKrishna committed:
Fix doc-string in serialization.py

Signed-off-by: Pradyumna Krishna <git@onpy.in>
